### ISSUE TITLE
Support for node_network_profile 

### ISSUE THREAD

**Issue by RelaxingLoki475** [Azure/terraform-azurerm-aks#524]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Have the ability to manage node_network_profile to set [application_security_group_ids](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#application_security_group_ids) in order to improve NSG management.

### New or Affected Resource(s)/Data Source(s)

azurerm_kubernetes_cluster

### References

https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#node_network_profile

**Comment by zioproto** [comment]
Documentation reference:
https://learn.microsoft.com/en-us/azure/aks/use-node-public-ips#use-public-ip-tags-on-node-public-ips-preview

**Comment by jramacpr** [comment]
I face the same issue when trying to attach an ASG with the node pool of the aks.Though terraform provides option to include the ASG inside the node network profile,it seems it do not work as expected.
  default_node_pool {
      name                   = var.agents_pool_name
      node_count             = var.agents_count
      vm_size                = var.agents_size
      vnet_subnet_id         = data.azurerm_subnet.subnet.id
      auto_scaling_enabled   = var.enable_auto_scaling
      max_count              = var.max_count
      min_count              = var.min_count
      node_public_ip_enabled = var.enable_node_public_ip
      max_pods               = var.agents_max_pods
      node_network_profile {
          application_security_group_ids = [data.azurerm_application_security_group.vm_asg.id]
      }
      upgrade_settings {
        max_surge = "10%"
      }
  }
After applying this ,the default node pool's vm scale set has no ASG's associated with it .Any manual association of the ASG is getting reverted to the original state(state from terraform config ,which sees as NO ASG's to be attached to the vm scale set).So as a result currently the attaching the ASG with the node network profile is not working as expected

**Comment by jramacpr** [comment]
@zioproto  iam using azurerm_kubernetes_cluster and as per the document it says
A node_network_profile block supports the following:



**[application_security_group_ids](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#application_security_group_ids) - (Optional) A list of Application Security Group IDs which should be associated with this Node Pool.**

**Comment by zioproto** [comment]
@jramacpr I opened PR #598 because it seems the module did not support `node_network_profile` in the default pool.

However, if you are consuming the resource `azurerm_kubernetes_cluster` directly you should open an issue at https://github.com/hashicorp/terraform-provider-azurerm/issues because this could be an issue with the upstream provider.

What version of the Hashicorp azurerm provider are you using in your project ?

thanks

**Comment by fuog** [comment]
> I face the same issue when trying to attach an ASG with the node pool of the aks. Though terraform provides option to include the ASG inside the node network profile, it seems it do not work as expected.
> 
> ```
>   default_node_pool {
>       name                   = var.agents_pool_name
>       node_count             = var.agents_count
>       vm_size                = var.agents_size
>       vnet_subnet_id         = data.azurerm_subnet.subnet.id
>       auto_scaling_enabled   = var.enable_auto_scaling
>       max_count              = var.max_count
>       min_count              = var.min_count
>       node_public_ip_enabled = var.enable_node_public_ip
>       max_pods               = var.agents_max_pods
>       node_network_profile {
>           application_security_group_ids = [data.azurerm_application_security_group.vm_asg.id]
>       }
>       upgrade_settings {
>         max_surge = "10%"
>       }
>   }
> ```
> 
> After applying this ,the default node pool's vm scale set has no ASG's associated with it .Any manual association of the ASG is getting reverted to the original state(state from terraform config ,which sees as NO ASG's to be attached to the vm scale set).So as a result currently attaching the ASG with the node network profile is not working as expected

We have the same problem as you describe. Did you find any solution? is the provider the problem?

**Comment by ahmadank** [comment]
@fuog  @jramacpr I might be a little late but I figured it out. Seems like to set the asg on the cluster you need to setup [allowed_host_ports](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#allowed_host_ports-1). This isn't a Terraform constraint but rather a Microsoft limitation. If you try attaching the ASG via the Azure CLI, you'll encounter the following error: `--asg-ids must be used with --allowed-host-ports.`; Will open a PR to throw an error and one to add a node to documentation when I get the chance

**Comment by zioproto** [comment]
I believe the feature is correctly explained here:
https://learn.microsoft.com/en-us/azure/aks/use-node-public-ips#allow-host-port-connections-and-add-node-pools-to-application-security-groups

**Comment by ahmadank** [comment]
Yes, thats the link I got the details from. Just found it a little odd the tf doesnt throw a warning or that its documented anywhere
